File: haskell-server/src/EagerWords/Server/Domain/Play.hs
--
-- Copyright 2017-2020 Azad Bolour
-- Licensed under GNU Affero General Public License v3.0 -
--   https://github.com/azadbolour/eagerwords/blob/master/LICENSE.md
--

{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE NamedFieldPuns #-}
{-# LANGUAGE DisambiguateRecordFields #-}
{-# LANGUAGE RecordWildCards #-}

module EagerWords.Server.Domain.Play (
    Play(..)
  , BasePlay(..)
  , mkWordPlay
  , mkSwapPlay
)
where

import qualified Data.ByteString.Lazy.Char8 as BC

import GHC.Generics (Generic)
import Data.Aeson (FromJSON, ToJSON)
import qualified Data.Aeson as Aeson

import Bolour.Plane.Domain.Point (Point)
import EagerWords.Common.Domain.Piece (Piece)
import EagerWords.Common.Domain.PlayerType (PlayerType(..))
import EagerWords.Common.Domain.PlayPiece (PlayPiece)

data PlayType = WordPlayType | SwapPlayType
  deriving (Eq, Show, Generic)

instance FromJSON PlayType
instance ToJSON PlayType

data BasePlay = BasePlay {
    playType :: PlayType
  , playNumber :: Int
  , playerType :: PlayerType
  , scores :: [Int]
}
  deriving (Eq, Show, Generic)

instance FromJSON BasePlay
instance ToJSON BasePlay

-- | Representation of a single play.
data Play =
  WordPlay {
      basePlay :: BasePlay
    , playPieces :: [PlayPiece]
    , replacementPieces :: [Piece]
    , deadPoints :: [Point]
  }
  | SwapPlay {
      basePlay :: BasePlay
    , swappedPiece :: Piece
    , newPiece :: Piece
  }
  deriving (Eq, Show, Generic)

instance FromJSON Play
instance ToJSON Play

encode :: Play -> String
encode play = BC.unpack $ Aeson.encode play

decode :: String -> Maybe Play
decode encoded = Aeson.decode $ BC.pack encoded

mkWordPlay ::
     Int
  -> PlayerType
  -> [Int]
  -> [PlayPiece]
  -> [Piece]
  -> [Point]
  -> Play
mkWordPlay playNumber playerType scores =
  let basePlay = BasePlay WordPlayType playNumber playerType scores
  in WordPlay basePlay

mkSwapPlay ::
     Int
  -> PlayerType
  -> [Int]
  -> Piece
  -> Piece
  -> Play
mkSwapPlay playNumber playerType scores =
  let basePlay = BasePlay SwapPlayType playNumber playerType scores
  in SwapPlay basePlay

File: scala-server/app/com/bolour/eagerwords/common/domain/Play.scala
/*
 * Copyright 2017-2020 Azad Bolour
 * Licensed under GNU Affero General Public License v3.0 -
 *    https://github.com/azadbolour/eagerwords.com/blob/master/LICENSE.md
 */

package com.bolour.eagerwords.common.domain

import com.bolour.eagerwords.common.domain.PlayerType._
import com.bolour.eagerwords.common.domain.PlayType._
import com.bolour.plane.domain.Point

/**
  * Representation of a play in the game - a word play or a swap play.
  * Used to store plays as JSON strings in the database.
  */
sealed abstract class Play(playType: PlayType, playNumber: Int, playerType: PlayerType, scores: List[Int])

/**
  * Construct by using mkWordPlay to prevent error in play type.
  * Forced to add redundant tag to easily do json.
  */
case class WordPlay(playType: PlayType, playNumber: Int, playerType: PlayerType, scores: List[Int],
  playPieces: List[PlayPiece], replacementPieces: List[Piece], deadPoints: List[Point]) extends Play(playType, playNumber, playerType, scores)

/**
  * Construct by using mkSwapPlay to prevent error in play type.
  * Forced to add redundant tag to easily do json.
  */
case class SwapPlay(playType: PlayType, playNumber: Int, playerType: PlayerType, scores: List[Int],
  swappedPiece: Piece, newPiece: Piece) extends Play(playType, playNumber, playerType, scores)

object Play {
  val PlayTypeFieldName = "playType"

  def mkWordPlay(playNumber: Int, playerType: PlayerType, scores: List[Int],
    playPieces: List[PlayPiece], replacementPieces: List[Piece], deadPoints: List[Point]): Play =
    WordPlay(WordPlayType, playNumber, playerType, scores, playPieces, replacementPieces, deadPoints)

  def mkSwapPlay(playNumber: Int, playerType: PlayerType, scores: List[Int],
    swappedPiece: Piece, newPiece: Piece): Play =
    SwapPlay(SwapPlayType, playNumber, playerType, scores, swappedPiece, newPiece)
}
